File: app/src/main/java/com/arkadiusz/dayscounter/Fragments/PastFragment.java
package com.arkadiusz.dayscounter.Fragments;

import static android.content.Context.VIBRATOR_SERVICE;

import android.content.DialogInterface;
import android.content.DialogInterface.OnClickListener;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.os.Vibrator;
import android.preference.PreferenceManager;
import android.support.annotation.Nullable;
import android.support.v4.app.ActivityOptionsCompat;
import android.support.v4.app.Fragment;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AlertDialog.Builder;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import com.arkadiusz.dayscounter.Activities.AddActivity;
import com.arkadiusz.dayscounter.Activities.DetailActivity;
import com.arkadiusz.dayscounter.Activities.MainActivity;
import com.arkadiusz.dayscounter.Adapters.RecyclerViewAdapter;
import com.arkadiusz.dayscounter.Database.Event;
import com.arkadiusz.dayscounter.Model.Migration;
import com.arkadiusz.dayscounter.Model.RecyclerItemClickListener;
import com.arkadiusz.dayscounter.Model.RecyclerItemClickListener.OnItemClickListener;
import com.arkadiusz.dayscounter.R;
import io.realm.Realm;
import io.realm.Realm.Transaction;
import io.realm.RealmConfiguration;
import io.realm.RealmResults;
import io.realm.Sort;


public class PastFragment extends Fragment {

  Realm realm;
  RealmResults<Event> results;
  RecyclerItemClickListener mListener;
  MainActivity activity;
  private String[] options;
  private String sortType;
  RecyclerView recyclerView;
  RecyclerViewAdapter adapter;


  @Override
  public void onActivityCreated(@Nullable Bundle savedInstanceState) {
    super.onActivityCreated(savedInstanceState);
  }

  @Override
  public void onCreate(@Nullable Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    receiveSortType();
    setUpRealm();
    setUpOptions();
  }

  @Nullable
  @Override
  public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container,
      @Nullable Bundle savedInstanceState) {
    View view = inflater.inflate(R.layout.past_fragment_xml, container, false);
    recyclerView = (RecyclerView) view.findViewById(R.id.past_recycler_view);
    recyclerView.setLayoutManager(new LinearLayoutManager(getContext()));
    adapter = new RecyclerViewAdapter(getContext(), results);
    recyclerView.setHasFixedSize(true);
    recyclerView.setAdapter(adapter);
    recyclerView.addOnItemTouchListener(new RecyclerItemClickListener(getContext(), recyclerView,
        new OnItemClickListener() {
          @Override
          public void onItemClick(View view, int position) {
            int id = results.get(position).getId();
            Intent intent = new Intent(getContext(), DetailActivity.class);
            intent.putExtra("event_id", id);
            String transitionName = "transition_detail";
            ActivityOptionsCompat options =
                ActivityOptionsCompat.makeSceneTransitionAnimation(getActivity(),
                    view,   //
                    transitionName    //
                );

            startActivity(intent, options.toBundle());

          }

          @Override
          public void onItemLongClick(View view, final int position) {
            vibration();
            AlertDialog.Builder builder = new AlertDialog.Builder(getContext());
            builder.setTitle(getString(R.string.fragment_main_dialog_title));
            builder.setItems(options, new DialogInterface.OnClickListener() {
              @Override
              public void onClick(DialogInterface dialog, int which) {
                switch (which) {
                  case 0:
                    int id = results.get(position).getId();
                    Intent intent = new Intent(getContext(), AddActivity.class);
                    intent.putExtra("mode", "edit");
                    intent.putExtra("id", id);
                    startActivity(intent);
                    break;
                  case 1:
                    new Builder(getContext())
                        .setTitle(getString(R.string.fragment_delete_dialog_title))
                        .setMessage(getString(R.string.fragment_delete_dialog_question))
                        .setPositiveButton(android.R.string.yes, new OnClickListener() {
                          public void onClick(DialogInterface dialog, int which) {
                            int id = results.get(position).getId();
                            final RealmResults<Event> results = realm.where(Event.class)
                                .equalTo("id", id).findAll();
                            realm.executeTransaction(new Transaction() {
                              @Override
                              public void execute(Realm realm) {
                                results.deleteAllFromRealm();
                              }
                            });

                          }
                        })
                        .setNegativeButton(android.R.string.no, new OnClickListener() {
                          public void onClick(DialogInterface dialog, int which) {
                            // do nothing
                          }
                        })
                        .show();
                    break;
                }
              }
            });
            builder.show();
          }
        }));

    return view;
  }


  public void vibration() {
    Vibrator v = (Vibrator) getContext().getSystemService(VIBRATOR_SERVICE);
    v.vibrate(25);
  }

  public void setUpRealm() {
    Realm.init(getContext());
    RealmConfiguration config = new RealmConfiguration.Builder()
        .schemaVersion(2)
        .migration(new Migration())
        .build();
    realm = Realm.getInstance(config);
    results = realm.where(Event.class).equalTo("type", "past").findAll();
    switch (sortType) {
      case "date_desc":
        results = results.sort("date", Sort.ASCENDING);
        break;
      case "date_asc":
        results = results.sort("date",Sort.DESCENDING);
        break;
      case "date_order":
        break;
    }
  }

  public void setUpOptions() {
    options = new String[2];
    options[0] = getString(R.string.fragment_main_dialog_option_edit);
    options[1] = getString(R.string.fragment_main_dialog_option_delete);
  }

  public void receiveSortType() {
    SharedPreferences sharedPreferences = PreferenceManager.getDefaultSharedPreferences(getActivity());
    sortType = sharedPreferences.getString("sort_type", "date_order");
  }

  public void refreshData() {
    receiveSortType();
    results = realm.where(Event.class).equalTo("type", "past").findAll();
    switch (sortType) {
      case "date_desc":
        results = results.sort("date", Sort.ASCENDING);
        break;
      case "date_asc":
        results = results.sort("date",Sort.DESCENDING);
        break;
      case "date_order":
        break;
    }
    adapter = new RecyclerViewAdapter(getContext(), results);
    recyclerView.setAdapter(adapter);
  }

}

File: app/src/main/java/com/arkadiusz/dayscounter/fragments/PastFragment.kt
package com.arkadiusz.dayscounter.fragments

import PreferenceUtils.defaultPrefs
import PreferenceUtils.get
import android.content.Context
import android.os.Build
import android.os.Bundle
import android.os.VibrationEffect
import android.os.Vibrator
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.arkadiusz.dayscounter.R
import com.arkadiusz.dayscounter.activities.DetailActivity
import com.arkadiusz.dayscounter.activities.EditActivity
import com.arkadiusz.dayscounter.adapters.EventsAdapter
import com.arkadiusz.dayscounter.adapters.RecyclerItemClickListener
import com.arkadiusz.dayscounter.model.Event
import com.arkadiusz.dayscounter.repositories.DatabaseProvider
import com.arkadiusz.dayscounter.repositories.FirebaseRepository
import com.arkadiusz.dayscounter.utils.RemindersUtils.deleteReminder
import io.realm.RealmResults
import kotlinx.android.synthetic.main.activity_main.*
import org.jetbrains.anko.alert
import org.jetbrains.anko.selector
import org.jetbrains.anko.startActivity

/**
 * Created by arkadiusz on 24.03.18
 */

class PastFragment : Fragment() {

    private val databaseRepository = DatabaseProvider.provideRepository()
    private val firebaseRepository = FirebaseRepository()

    private var sortType = ""
    private lateinit var adapter: EventsAdapter
    private lateinit var recyclerView: androidx.recyclerview.widget.RecyclerView
    private lateinit var eventsList: RealmResults<Event>
    private lateinit var eventContextOptions: List<String>

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        receiveSortType()
        setUpContextOptions()
        setUpData()
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(R.layout.future_fragment, container, false)
        initRecyclerView(view)
        setUpRecyclerView()
        hideFABOnScroll()
        return view
    }

    private fun initRecyclerView(view: View) {
        recyclerView = view.findViewById(R.id.recyclerView)
        recyclerView.layoutManager = LinearLayoutManager(context)
        recyclerView.setHasFixedSize(true)
        recyclerView.addOnItemTouchListener(object : RecyclerItemClickListener(context!!, recyclerView, object : OnItemClickListener {
            override fun onItemClick(view: View?, position: Int) {
                val id = eventsList[position].id
                context?.startActivity<DetailActivity>("event_id" to id)
            }

            override fun onItemLongClick(view: View?, position: Int) {
                vibration()
                displayEventOptions(eventsList[position].id, eventsList[position])
            }

        }) {})
    }

    private fun receiveSortType() {
        context?.let {
            val sharedPref = defaultPrefs(it)
            sortType = sharedPref["sort_type"] ?: "date_order"
        }
    }

    private fun setUpContextOptions() {
        eventContextOptions = listOf(getString(R.string.fragment_main_dialog_option_edit),
                getString(R.string.fragment_main_dialog_option_delete))
    }

    private fun setUpData() {
        eventsList = databaseRepository.getPastEvents()
        sortEventsList()
    }

    private fun sortEventsList() {
        when (sortType) {
            "date_desc" -> eventsList = databaseRepository.sortEventsDateAsc(eventsList)
            "date_asc" -> eventsList = databaseRepository.sortEventsDateDesc(eventsList)
        }
    }

    private fun setUpRecyclerView() {
        adapter = EventsAdapter(context!!, eventsList)
        recyclerView.adapter = adapter
        recyclerView.scheduleLayoutAnimation()
        recyclerView.invalidate()
    }

    private fun vibration() {
        val vibrator = context?.getSystemService(Context.VIBRATOR_SERVICE) as Vibrator
        if (Build.VERSION.SDK_INT >= 26) {
            vibrator.vibrate(VibrationEffect.createOneShot(25, VibrationEffect.DEFAULT_AMPLITUDE))
        } else {
            vibrator.vibrate(25)
        }
    }

    private fun displayEventOptions(eventId: Int, event: Event) {
        context?.let { ctx ->
            ctx.selector(getString(R.string.fragment_main_dialog_title), eventContextOptions) { _, i ->
                when (i) {
                    0 -> ctx.startActivity<EditActivity>("eventId" to eventId)
                    1 -> {
                        ctx.alert(getString(R.string.fragment_delete_dialog_question)) {
                            positiveButton(android.R.string.yes) {
                                deleteReminder(context!!, event)
                                databaseRepository.deleteEventFromDatabase(eventId)
                                firebaseRepository.deleteEvent(defaultPrefs(context!!)["firebase-email"]
                                        ?: "", eventId)
                            }
                            negativeButton(android.R.string.no) {}
                        }.show()
                    }
                }
            }
        }
    }

    private fun hideFABOnScroll() {
        recyclerView.addOnScrollListener(object : RecyclerView.OnScrollListener() {
            override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
                if (dy > 0) {
                    activity?.fab?.hide()
                } else if (dy < 0) {
                    activity?.fab?.show()
                }
            }
        })
    }

    fun refreshData() {
        receiveSortType()
        setUpData()
        setUpRecyclerView()
    }

}